packages/pubsub/src/queue.js

```diff
@@ -0,0 +1,200 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module pubsub/queue
+ */
+
+'use strict';
+
+var common = require('@google-cloud/common');
+var prop = require('propprop');
+
+/**
+ * Semi-generic Queue object used to queue up multiple requests while limiting
+ * the number of concurrent requests.
+ *
+ * @constructor
+ * @alias module:pubsub/queue
+ *
+ * @param {object} options - Queue configuration settings.
+ * @param {number} options.maxBytes - Max size of data to be sent per request.
+ * @param {number} options.maxMessages - Max number of messages to be sent per
+ *     request.
+ * @param {number} options.maxMilliseconds - Max time to wait before sending
+ *     data.
+ * @param {number} options.maxRequests - Max number of concurrent requests
+ *     allowed.
+ * @param {function(messages, callback)} options.sendHandler - Function to be
+ *     called with message queue.
+ *
+ * @example
+ * var queue = new Queue({
+ *   maxBytes: Math.pow(1024, 2) * 5,
+ *   maxMessages: 1000,
+ *   maxMilliseconds: 1000,
+ *   maxRequests: 5,
+ *   send: function(messages, done) {
+ *     var reqOpts = {
+ *       topic: self.topic.name,
+ *       messages = messages.map(function(message) {
+ *         return {
+ *           data: message.data,
+ *           attributes: message.attrs
+ *         };
+ *       })
+ *     };
+ *
+ *     return self.api.Publisher.publish(reqOpts, function(err, resp) {
+ *       if (err) {
+ *         done(err);
+ *         return;
+ *       }
+ *
+ *       done(null, {
+ *         responses: resp.messageIds
+ *       })
+ *     });
+ *   }
+ * });
+ */
+function Queue(options) {
+  this.maxBytes = options.maxBytes || Math.pow(1024, 2) * 5;
+  this.maxMessages = options.maxMessages || 1000;
+  this.maxMilliseconds = options.maxMilliseconds || 1000;
+  this.sendHandler = options.send;
+
+  this.inventory = {
+    queued: new Set(),
+    inFlight: new Set()
+  };
+
+  Object.defineProperty(this.inventory, 'queueBytes', {
+    get: function() {
+      var size = 0;
+
+      this.queued.forEach(function(message) {
+        size += message.data.size;
+      });
+
+      return size;
+    }
+  });
+
+  this.intervalHandle_ = null;
+}
+
+/**
+ * Adds message to the queue.
+ *
+ * @param {*} data - The message data.
+ * @param {function} callback - Callback to be ran once message is sent.
+ *
+ * @example
+ * queue.add(1234, function(err, resp) {});
+ */
+Queue.prototype.add = function(data, callback) {
+  var hasMaxMessages = (this.inventory.queued.size + 1) >= this.maxMessages;
+  var hasMaxBytes = (this.inventory.queueBytes + data.size) >= this.maxBytes;
```

Comment:
I am not particularly worried about this one. The blunt issue here is that we can only estimate the size of a message anyway (we do not have a good way to know exactly what the size of a serialized message is). Since we do not have a sufficiently precise understanding of how many bytes we _have_ thus far, being off by one byte is a pretty meaningless distinction here.